Merkle Tree construction from elements

lazcanoluca opened the pull request:
Provides the base structure and implementation of the Merkle Tree. Basic methods for hashing and computation of the Merkle root have been implemented and tested. This PR provides the ability to create a new Merkle Tree from a list of elements. The resulting Merkle Tree contains the leaves and the Merkle root.

lazcanoluca committed:
Add features roadmap to README

lazcanoluca committed:
Fix features checkboxes README

lazcanoluca committed:
Remove `lib` default contennt

lazcanoluca committed:
Add MerkleTree module

lazcanoluca committed:
Implement the base hashing function with SHA-256

lazcanoluca committed:
Add method to get the hash of left and right hashes

lazcanoluca committed:
Add method to compute the next level of the tree

lazcanoluca committed:
Implement method to compute the merkle root

lazcanoluca committed:
Complete `MerkleTree` struct with root and leafs

lazcanoluca committed:
Implement `build` method

This method is the entry point for creating a `MerkleTree` with several items.

lazcanoluca committed:
Update README

juanimedone commented on src/merkle_tree.rs:
```diff
@@ -0,0 +1,221 @@
+
+pub struct MerkleTree {
+    pub root: Option<[u8; 32]>,
+    pub leafs: Vec<[u8; 32]>,
```
Just a typo, it should be `leaves` instead of `leafs`.

lazcanoluca replied:
yikes, you're right.

juanimedone commented on src/merkle_tree.rs:
```diff
@@ -0,0 +1,221 @@
+}
+
+// TODO: type Hash for [u8; 32]
+// TODO: type Level for Vec<[u8; 32]> of even length (duplicates last element if odd)
```
What would that be for?

lazcanoluca replied:
I like having clear types with descriptive names, even if they just "wrap" a simple type. I feel it makes the code clearer to read.

I didn't mean to include the TODO's in the commit tho, they were just ideas for making it better in the future.

juanimedone commented on src/merkle_tree.rs:
```diff
@@ -0,0 +1,221 @@
+    /// Create a new MerkleTree from the provided items.
+    /// Each item should be representable as bytes.
+    /// It returns a `MerkleTree` instance with the leaf hashes and the Merkle root.
+    pub fn build<T: AsRef<[u8]>>(&self, items: &[T]) -> Self {
+        let leafs: Vec<[u8; 32]> = items.iter().map(|item| self.hash(item.as_ref())).collect();
+
+        Self {
+            root: Some(self.merkle_root(leafs.clone())),
+            leafs,
+        }
+    }
```
This function should probably check that `items` is not empty.

lazcanoluca replied:
So true!

Do you think it should `Err` if there were no items? Or just return an empty tree?

juanimedone replied:
Given that the tree will have the feature of dynamically adding elements I think the second option would be better.

juanimedone reviewed (changes requested):
The code looks very nice overall, and it's a great starting point. Implementing a generic `build` method is an excellent idea. I'm just leaving a few comments with details and a question.

PS: Remember to run `cargo clippy` for linting and to catch any potential issues.

Oppen commented on src/merkle_tree.rs:
```diff
@@ -0,0 +1,221 @@
+    pub leafs: Vec<[u8; 32]>,
+}
+
+impl Default for MerkleTree {
+    fn default() -> Self {
+        Self {
+            root: None,
+            leafs: Vec::new(),
+        }
+    }
+}
```
I'm not sure a `default` implementation makes sense for `MerkleTree`.

lazcanoluca replied:
Okay! I'm not really sure why I included it originally.

lazcanoluca replied:
I remember now why I added the `Default` trait! It was suggested to me by clippy:

```
% cargo clippy
    Checking merkle-tree v0.1.0 (/Users/luca/Documents/merkle-tree)
warning: you should consider adding a `Default` implementation for `MerkleTree`
  --> src/merkle_tree.rs:10:5
   |
10 | /     pub fn new() -> Self {
11 | |         Self {
12 | |             root: None,
13 | |             leaves: Vec::new(),
14 | |         }
15 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#new_without_default
   = note: `#[warn(clippy::new_without_default)]` on by default
help: try adding this
   |
8  + impl Default for MerkleTree {
9  +     fn default() -> Self {
10 +         Self::new()
11 +     }
12 + }
   |

warning: `merkle-tree` (lib) generated 1 warning (run `cargo clippy --fix --lib -p merkle-tree` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.14s
```

Oppen replied:
I would say an empty `new` doesn't make much sense either.

lazcanoluca replied:
I see, I think you're right.

lazcanoluca committed:
Fix typo leafs

lazcanoluca committed:
Remove unnecesary TODOs

lazcanoluca committed:
Remove Default implementation

lazcanoluca committed:
Refactor methods to be static

The `build` method is changed to be static, so a new Merkle Tree can be directly created from the provided elements. Before, the `build` function acted on an existing (empty) `MerkleTree`, which I consider to be unnecesary overhead.
Furthermore, other private methods that act as "utils" (hashing, tree level computation) have also been made static, as they do not act on the Merkle Tree data structure itself, only on the method inputs.

lazcanoluca committed:
Introduce condition for `build` with no items

Oppen commented on src/merkle_tree.rs:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    /// Computes the parent hash for the concatenation of the provided left and right hashes.
+    fn merkle_parent(left: &[u8; 32], right: &[u8; 32]) -> [u8; 32] {
+        let concat = [left.as_slice(), right.as_slice()].concat();
+
+        Self::hash(&concat)
+    }
```
`left` and `right` are always consecutive in memory, so we may save ourselves an allocation by using [as_flattened](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_flattened):
```suggestion
    fn merkle_parent(children: &[[u8; 32]; 2]) -> [u8; 32] {
        Self::hash(children.as_flattened())
    }
```

lazcanoluca replied:
Based on your suggestion, I propose this:

```
    /// Computes the parent hash for the concatenation of the children hashes.
    fn merkle_parent(children: &[[u8; 32]]) -> [u8; 32] {
        Self::hash(children.as_flattened())
    }
```

for two reasons:
1. The method is easier to use, I don't have to check the length of the `children` slice equals 2 each time I want to use it.
2. It actually makes the method more generic, as it's not tied to having two children. It can be used to compute the merkle parent of an arbitrary amount of children hashes.

What do you think?

Oppen commented on src/merkle_tree.rs:
```diff
@@ -0,0 +1,206 @@
+use hmac_sha256::Hash;
+
+pub struct MerkleTree {
+    pub root: Option<[u8; 32]>,
+    pub leaves: Vec<[u8; 32]>,
+}
```
Is there any reason you chose not to store the intermediate branches? If so, the rationale and tradeoffs made should be documented either here or in the README, or both.

lazcanoluca replied:
If I understand correctly, the leaves (given the same ordering) are enough to compute the Merkle Tree. Storing the intermediate nodes is more of an optimization to avoid recomputing the tree each time it is queried. I don't rule out storing the intermediate nodes in the future, I just don't think it is necessary right now.

Do you agree?

Oppen commented on src/merkle_tree.rs:
```diff
@@ -0,0 +1,206 @@
+
+impl MerkleTree {
+    /// Create a new empty MerkleTree instance.
+    pub fn new() -> Self {
+        Self {
+            root: None,
+            leaves: Vec::new(),
+        }
+    }
+
+    /// Create a new MerkleTree from the provided items.
+    /// Each item should be representable as bytes.
+    /// It returns a `MerkleTree` instance with the leaf hashes and the Merkle root.
+    pub fn build<T: AsRef<[u8]>>(items: &[T]) -> Self {
+        if items.is_empty() {
+            return Self::new();
+        }
+
+        let leaves: Vec<[u8; 32]> = items.iter().map(|item| Self::hash(item.as_ref())).collect();
+
+        Self {
+            root: Some(Self::merkle_root(leaves.clone())),
+            leaves,
+        }
+    }
```
I'd like to have someone else's output on this, but in my head rather than having an empty tree case, I'd have all trees having something inside, and `build` return an `Option<MerkleTree>`, where passing an empty slice produces `None`.

lazcanoluca replied:
We talked a bit about this [here](https://github.com/lazcanoluca/merkle-tree/pull/1#discussion_r1694438280), where we concluded it would be okay to build an empty Merkle Tree from an empty slice. I agree using an `Option` makes total sense. I think I'm going with the `Option`.

lazcanoluca committed:
Change `merkle_parent` arguments

Instead of having a `left` and `right` children and concatenating them, the method now receives a slice with all the children to hash the parent.

lazcanoluca committed:
Change `build` signature so it returns an `Option`

Trying to build a `MerkleTree` with no items returns now returns `None`.

lazcanoluca committed:
Remove `new` for `MerkleTree`

lazcanoluca committed:
Simplify redudant closure
